feature(search) - Allow aliases for durations

wmak opened the pull request:
- This allows us to use `ms`, `s`, `min`, `hr`, `day` and `wk` in our searches
- These values were selected to match what we display on the frontend,
  which is why its different from our other durations eg. `hr` instead of
  `h`

wmak committed:
feature(search) - Allow aliases for durations

- This allows us to use ms, s, hr, day and week in our searches
- These values were selected to match what we display on the frontend,
  which is why its different from our other durations `hr` instead of
  `h`

evanh reviewed (approved):
LGTM

markstory commented on src/sentry/api/event_search.py:
```diff
@@ -421,6 +430,20 @@ def visit_time_filter(self, node, children):
+            try:
+                search_value = parse_duration(*search_value.match.groups())
+            except InvalidQuery as exc:
+                raise InvalidSearchQuery(six.text_type(exc))
```
It would be good to have some tests around these failure scenarios as we shouldn't get 500s from them.

wmak committed:
ref: Adding tests for invalid durations

wmak committed:
fixup: - Adding invalid float error and catching it for aggregates

markstory commented on src/sentry/api/event_search.py:
```diff
@@ -229,6 +233,7 @@ class SearchVisitor(NodeVisitor):
     # A list of mappers that map source keys to a target name. Format is
     # <target_name>: [<list of source names>],
     key_mappings = {}
+    duration_keys = set(["transaction.duration", "p75", "p95", "p99"])
```
How do you see this working with functions that sometimes use duration and sometimes don't? For example `avg(transaction.duration)`

wmak replied:
🤔 What are the cases where `avg(transaction.duration)` won't use duration?

markstory replied:
You could give it `stack.colno`

wmak replied:
Ah hm, someone could do `avg(stack.colno):>500s` and it would become `avg(stack.colno):>500000` which would be confusing (both in why it changes and why a user entered a column number greater than 500 seconds)
I'll look into adding something that will parse the column out of the function and check the `duration_keys`

evanh replied:
I added a `resolve_function` in event_search that does all the column parsing and column validation, so I would update that to validate the column passed in.

evanh replied:
`event_search` no longer has the pXX aliases anywhere, so these should be removed.

wmak committed:
ref: Only apply duration parsing to durations

- Check that the column inside an aggregate is a duration before we
  parse the value as a duration

evanh commented on src/sentry/api/event_search.py:
```diff
@@ -389,12 +394,26 @@ def visit_numeric_filter(self, node, children):
+                _, agg_additions = resolve_field(search_key.name, None)
+                if len(agg_additions) > 0:
+                    # Extract column and function name out so we can check if we should parse as duration
+                    column, function = agg_additions[0][-2:]
```
You don't need to check the function here. Just the column. The only functions that don't have the column correctly will be apdex and impact and you can't filter those by duration anyways. Also the only value for `duration_keys` at this point is `transaction.duration`. The pXX aliases will be returned as functions here.

evanh commented on tests/sentry/snuba/test_discover.py:
```diff
@@ -903,6 +903,44 @@ def test_alias_aggregate_conditions(self, mock_query):
+                filter_keys={"project_id": [self.project.id]},
+                groupby=["transaction"],
+                dataset=Dataset.Discover,
+                aggregations=[["quantile(0.95)(duration)", None, "p95"]],
```
```suggestion
                aggregations=[["quantile(0.95)", "duration", "percentile_transaction_duration_0_95"]],
```
I think the having check will also fail.

evanh commented on tests/sentry/api/test_event_search.py:
```diff
@@ -549,6 +550,32 @@ def test_invalid_numeric_fields(self):
+        with self.assertRaises(InvalidSearchQuery, expected_regex="not a valid duration value"):
+            parse_search_query("transaction.duration:>..500s")
+
+    def test_invalid_aggregate_duration_filter(self):
```
Should also add a test for the wrong column case.

markstory commented on src/sentry/search/utils.py:
```diff
@@ -40,6 +40,32 @@ def parse_status_value(value):
+        delta = timedelta(hours=value)
+    elif interval == "day":
+        delta = timedelta(days=value)
+    elif interval == "wk":
```
Based on Adhiraj's feedback at weekly review, it would be good to also support the same duration period names as issue search does. So `w` `h` `d` etc. Having both the issue form and discover share code as well.

wmak committed:
ref - Addressing the PR comments

- Adding single letter aliases for the durations
- Removing the pXX function names
